qualtran/_infra/registers.py

```diff
@@ -61,24 +61,27 @@ class Register:
     """
 
     name: str
-    _bitsize: Union[int, QDType] = field(
-        converter=lambda v: v if isinstance(v, QDType) else QBit() if v == 1 else QAny(v)
-    )
+    # _bitsize: Union[int, QDType] = field(
+    #     converter=lambda v: v if isinstance(v, QDType) else QBit() if v == 1 else QAny(v)
+    # )
+    dtype: QDType
     shape: Tuple[int, ...] = field(
         default=tuple(), converter=lambda v: (v,) if isinstance(v, int) else tuple(v)
     )
     side: Side = Side.THRU
 
     def __attrs_post_init__(self):
-        if isinstance(self._bitsize, BoundedQUInt):
+        if not isinstance(self.dtype, QDType):
+            raise ValueError(f'dtype must be a QDType: found {type(self.dtype)}')
+        if isinstance(self.dtype, BoundedQUInt):
```

Comment:
I missed this from before. Why is this restriction in place? In principle you could have an array of numbers each from e.g. 1 to 5.

If we do want this: maybe make more general by having `QDType.validate_shape(shape)` method if there's some fundamental reason why some types shouldn't be arrayable

Reply:
This was copied over from a similar check in  SelectionRegister. I can't remember if it was a fundamental limitation there or just in place to prevent some untested functionality.

Reply:
@tanujkhattar I remember asking before about why multidimensional selection registers were frowned upon, but can't quite remember the conclusion.

Reply:
`Signature([Register('x', BoundedQUInt(n, N)), Register('y', BoundedQUint(m, M)])` represents signature of a Bloq that expects to do a coherent for-loop over `N * M` values; where variable `x` can take values between `[0, N)` and variable `y` can take values between `[0, M)`. 

`Signature([Register('x', BoundedQUInt(n, N), shape=(2, )])` - Now this represents two variables `x[0]` and `x[1]`, both of which can take values between `[0, N)`. So a total of `N ** 2` values. 

As the size of the multidimensional register increases, the number of values we can iterate upon increases exponentially. Also, the iteration length needs to be same across all variables. And it's a bit nontrivial to reason about the correctness of this case. For all these reasons, we decided to avoid multi-dimensional selection registers and forced users to declare one selection registers for every for-loop variable as part of the signature of the Bloq; even if two of the variables had the same iteration length and could in principle be represented using a multi dimensional selection register. 


Tl;Dr - It was mainly to enforce more expressive workflow and avoid confusing constructs in the specific case of selection registers. I think `BoundedQUint` is a more general abstraction that I can imagine using outside of the selection registers context (eg: to represent integers mod M) and so I think we can drop this assertion.

Reply:
I'll remove the exception.